### PR TITLE
JLP-87: Recover an access and safeguard tokens.

### LINKED ISSUE #87: Recover an access and safeguard tokens.
As a user that has lost his only device, I want to recover access to my wallet and safeguard my tokens.

- Setup a user wallet with a single device.
- Authorize a session key on the device.
- Initiate a recovery process for the user wallet (by assuming the only device is lost) and wait it to succeed.
- Logout previously authorized session keys.
- Authorize a new session key on the new device.

### PR DESCRIPTION
**Pull request by pgev**
fixes #87

### CODE REVIEW

File: test/run_tests.js [hunk 1]
```diff
@@ -80,6 +80,11 @@ const loadMocha = () => {
       // Skipping the file that manages shared state across tests.
       file => file !== path.join(__dirname, 'shared.js'),
     )
+    .filter(
+      file => file.indexOf(
```
Comment (jasonklein) [inline test/run_tests.js]: Please add a comment explaining what is being filtered, consistent with the other filters.

File: test/run_tests.js [hunk 2]
```diff
@@ -80,6 +80,11 @@ const loadMocha = () => {
...
+        'test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/' === -1,
```
Comment (jasonklein) [inline test/run_tests.js]: I assume this is meant to filter out the helper files added in this PR, but I don't think it is filtering anything. Maybe the closing parenthesis on L86 is wrongly-placed?

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
+'use strict';
+
+const { ContractInteract, Helpers } = require('@openst/openst.js');
+
+const UserWallet = require('./user_wallet');
+// const OpenST = require('../../../../../src/openst');
```
Comment (jasonklein) [inline test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js]: Please resolve.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+   * @returns {Object} Returns newly created UserWallet object.
+   *
+   * @memberof UserWalletFactory
+   */
+  async create(
+    gnosisSafeArgs,
+    delayedRecoveryModuleArgs,
+    tokenHolderArgs,
+  ) {
+    // @todo Assert that gnosisSafeOwnerKey.address is in gnosisSafeArgs.owners list.
```
Comment (jasonklein) [inline test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js]: Please resolve.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/recovery_controller.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      oldOwnerAddress,
+    );
+
+    await this.userWallet.delayedRecoveryModule.initiateRecovery(
+      prevOwnerAddress, oldOwnerAddress, newOwnerAddress,
+      signature.r, signature.s, signature.v,
+      { from: this.recoveryControllerAddress },
+    );
+
+    return this.userWallet.delayedRecoveryModule.auxiliaryWeb3.eth.getBlockNumber();
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/recovery_controller.js]: Indeed. Fixed.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    );
+
+    userWalletOwner = new UserWalletOwner(ownerKey, userWallet);
+
+    recoveryOwner = new RecoveryOwner(recoveryOwnerKey, userWallet);
+
+    recoveryController = new RecoveryController(recoveryControllerAddress, userWallet);
+  });
+
+  it('Authorize a session.', async () => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: The issue is a user story related to delayed recovery module. Please, let's have a look into ticket description and see can we improve it.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    await recoveryController.executeRecovery(
+      userWalletOwner.ownerKey.address,
+      newOwnerKey.address,
+    );
+
+    userWalletNewOwner = new UserWalletOwner(newOwnerKey, userWallet);
+  });
+
+
+  it('Logout all previously authorized session keys.', async () => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: Same comment as in elsewhere for now.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+      sessionKey.address,
+      sessionKeySpendingLimit,
+      sessionKeyExpirationHeight,
+      [signature],
+    );
+  });
+
+  let newOwnerKey = '';
+
+  let recoveryInitiationBlockNumber = -1;
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: Exactly.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    const web3 = connection.auxiliaryWeb3;
+
+    newOwnerKey = await web3.eth.accounts.create();
+
+    const signature = await recoveryOwner.signInitiateRecovery(
+      userWalletOwner.ownerKey.address,
+      newOwnerKey.address,
+    );
+
+    recoveryInitiationBlockNumber = await recoveryController.initiateRecovery(
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: I've forgot about that interface :). Thx. Fixed.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+  });
+
+  let newOwnerKey = '';
+
+  let recoveryInitiationBlockNumber = -1;
+
+  it('Initiate a recovery process to replace the lost device.', async () => {
+    const web3 = connection.auxiliaryWeb3;
+
+    newOwnerKey = await web3.eth.accounts.create();
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: I would rather not use mock data in a places that can be avoided.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+  });
+
+  let userWalletNewOwner = {};
+
+  it('Execute recovery process to replace the lost device.', async () => {
+    const web3 = connection.auxiliaryWeb3;
+
+    await waitBlockNumber(web3, recoveryInitiationBlockNumber + recoveryBlockDelay);
+
+    await recoveryController.executeRecovery(
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: I've added doc comment to the RecoveryController::initiateRecovery and RecoveryController::executeRecovery functions.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+const UserWalletFactory = require('./helpers/user_wallet_factory');
+const UserWalletOwner = require('./helpers/user_wallet_owner');
+const RecoveryOwner = require('./helpers/recovery_owner');
+const RecoveryController = require('./helpers/recovery_controller');
+
+function sleep(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function waitBlockNumber(web3, blockNumber) {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: I've created test/utils.js and moved the functions there. Thx for heads up.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    currentBlockNumber = await web3.eth.getBlockNumber();
+  }
+}
+
+describe('Recover an access', async () => {
+  let chainConfig = {};
+  let connection = {};
+  let openst = {};
+
+  // @todo Who/where should deploy organization and economy token contracts.
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: I would rather remove this todo for now and reevaluate it later if needed. Thx for heads up.

File: test/run_tests.js (hunk 1 above)
Comment (pgev) [reply test/run_tests.js]: Done.

File: test/run_tests.js (hunk 2 above)
Comment (pgev) [reply test/run_tests.js]: Yes. Wrongly placed. Fixed (thx!).

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js (hunk 3 above)
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js]: Done.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js (hunk 4 above)
Comment (pgev) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/helpers/user_wallet_factory.js]: Done.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jasonklein) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: @pgev: I see that about the user story and ticket, but the ticket is silent on the question of organization/architecture. Or is co-location of responsive components a standard operating procedure for implementing user stories? I can see that being true given that the aim should be fairly scoped.

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (jasonklein) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: An even better solution. Thanks!

File: test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (jasonklein) [reply test/04_openst/gnosisSafeModules/delayedRecoveryModule/initiate_and_execute_recovery.js]: A better solution. 🌟

**Pull request merged**